pkg/reconciler/shared/tektonconfig/pipeline/pipeline.go

```diff
@@ -20,35 +20,51 @@ import (
 	"context"
 	"errors"
 	"fmt"
-	"log"
 	"reflect"
 
 	"github.com/tektoncd/operator/pkg/apis/operator/v1alpha1"
 
 	op "github.com/tektoncd/operator/pkg/client/clientset/versioned/typed/operator/v1alpha1"
-	operatorv1alpha1 "github.com/tektoncd/operator/pkg/client/clientset/versioned/typed/operator/v1alpha1"
 	"github.com/tektoncd/operator/pkg/reconciler/common"
 	apierrs "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/wait"
-	"knative.dev/pkg/test/logging"
 )
 
-func CreatePipelineCR(ctx context.Context, instance v1alpha1.TektonComponent, client operatorv1alpha1.OperatorV1alpha1Interface) error {
-	configInstance := instance.(*v1alpha1.TektonConfig)
-	if _, err := ensureTektonPipelineExists(ctx, client.TektonPipelines(), configInstance); err != nil {
-		return errors.New(err.Error())
+func EnsureTektonPipelineExists(ctx context.Context, clients op.TektonPipelineInterface, config *v1alpha1.TektonConfig) (*v1alpha1.TektonPipeline, error) {
+	tpCR, err := GetPipeline(ctx, clients, v1alpha1.PipelineResourceName)
+
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			tpCR = &v1alpha1.TektonPipeline{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: v1alpha1.PipelineResourceName,
+				},
+				Spec: v1alpha1.TektonPipelineSpec{
+					CommonSpec: v1alpha1.CommonSpec{
+						TargetNamespace: config.Spec.TargetNamespace,
+					},
+					Pipeline: config.Spec.Pipeline,
+					Config:   config.Spec.Config,
+				},
+			}
+			_, err = clients.Create(ctx, tpCR, metav1.CreateOptions{})
+			if err != nil {
+				return nil, err
+			}
```

Comment:
```suggestion
			tp, err = clients.Create(ctx, tpCR, metav1.CreateOptions{})
			if err != nil {
				return nil, err
			}
			return tp, v1alpha1.RECONCILE_AGAIN_ERR
```

if err is nil, we return `RECONCILE_AGAIN_ERR` 

1.  check a small piece of state (eg: does pipeline cr exists)
     1.a if the check is false, take smallest unit of action towards reaching the checked state
     1.b  return and try the check in step 1
2. check next small piece of state (eg: is pipeline cr valid or ready)
...

Reply:
in other words, the reconciler should try `ok, err := isTektonPipelineReady(tpCR, err)` 
only if the previous check `tpCR, err := GetPipeline(ctx, clients, v1alpha1.PipelineResourceName)` was successful in the same loop.

note: This is pattern  is not just something that we need to follow in this specific function or this operator. This is the core tenet of declarative coding paradigm in k8s. 🧑‍💻 


~~~

         |                                    |
         |                                    |
         |                                    |
       /   \                                  |
      /     \                                 | Requeue reconcile
     /       \   false ____________________   |
    / check 1 \________| Act on the cluster|__|
    \         /        |___________________|  |
     \       /                                |
       \   /                                  |
         |                                    |
         | true                               |
         |                                    |
       /   \                                  |
      /     \                                 | Requeue reconcile
     /       \   false ____________________   |
    / check 2 \________| Act on the cluster|__|
    \         /        |___________________|
     \       /
       \   /
         |
         | true
         |
         .
         .
         .
~~~

Reply:
yeah, the logic looks right, i was tripped by the `else` 🤓